
## Snakemake Workflow

Snakemake is a workflow management system that aims to make the definition and execution of data analysis workflows easy, reproducible, and scalable. It uses a Python-based language to specify workflows in terms of rules that describe how to derive output files from input files.
For a comprehensive introduction and installation of Snakemake, refer to the official tutorial: (https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html).


## Environment setup

## Example data

1. Create a directory for the task

   $ mkdir Snakemak

1. Download the new dataset from this link https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9205427/#sec4.3.2.
1. Synchronize the input files using `rsync`: To do this between your local machine and the remote server use the following command:

## Snakemake workflows 

### Reads alignment

### Converting sam file to bam file

### Sorting read alignments

### Indexing read alignment 

### Extracting chromosome6 

### Runing pbhoney for variant calling 

### Runing snifles for variant calling

