SAMPLE = ["A","B"]

rule all:
    input:"plots/quals.svg"

rule bwa_map:
    input:
        "data/genome.fa",
        "data/samples/{sample}.fastq"
    output:
        "output/mapped_reads/{sample}.bam"
    shell:
        "bwa mem {input} | samtools view -Sb - > {output}"
        
rule samtools_sort:
    input:
        "output/mapped_reads/{sample}.bam"
    output:
        "output/sorted_reads/{sample}.bam"
    shell:
       "samtools sort -T sorted_reads/{wildcards.sample} "
        "-O bam {input} > {output}"


rule samtools_index:
    input:
        "output/sorted_reads/{sample}.bam"
    output:
        "output/sorted_reads/{sample}.bam.bai"
    shell:
        "samtools index {input}"


rule bcftools_call: 
    input:
        fa="data/genome.fa",
        bam=expand("output/sorted_reads/{sample}.bam", sample=SAMPLE),
        bai=expand("outpu/sorted_reads/{sample}.bam.bai", sample=SAMPLE)
    output:
        "calls/all.vcf"
    shell:
        "bcftools mpileup -f {input.fa} {input.bam} | "
        "bcftools call -mv -> {output}"
rule plot_quals:
    input:
        "calls/all.vcf"
    output:
        "plots/quals.svg"
    script:
        "script/plot-quals.py"
